### PR TITLE
Improve performance of opnorm for (m x 1) and (1 x n) sparse matrices

### PR DESCRIPTION
**Pull request by goggle**
This PR improves the performance of `opnorm` for `m x 1` and `1 x n` sparse matrices by using `nzval` directly instead of allocating a dense array first.

To test the performance improvement, I have created two sparse matrices `A` and `B` with `A = sprandn(100_000, 1, 0.6)`, `B = sprandn(1, 100_000, 0.6)` and benchmarked `opnorm` using `BenchmarkTools.jl`. Here are the results:

| command        | Time (master) | Time (PR)   | Allocations (master)  | Allocations (PR)
| ------------- |-------------:| -----:|------:|-------:|
| `opnorm(A, 1)`  | 239.191 μs | 9.361 μs | 781.34 KiB | 16 bytes  |
| `opnorm(A)`  | 188.754 μs | 15.220 μs | 781.34 KiB | 16 bytes  |
| `opnorm(A, Inf)`  | 350.948 μs | 83.312 μs | 781.34 KiB | 16 bytes  |
| `opnorm(B, 1)`  |834.196 μs | 83.252 μs | 781.34 KiB | 16 bytes  |
| `opnorm(B)`  | 675.194 μs | 15.216 μs | 781.34 KiB  | 16 bytes |
| `opnorm(B, Inf)`  | 723.402 μs |9.448 μs | 781.34 KiB | 16 bytes  |

There could be one issue: Resizing `A.nzval` and changing a value of `A.nzval` outside of the array's size does affect the result of `opnorm(A, p)`. If that should not happen, this PR needs some rework. In that case, I would be thankful for some tips.

### CODE REVIEW

File: stdlib/SparseArrays/src/linalg.jl [hunk 1]
```diff
@@ -1028,9 +1028,16 @@ function opnorm(A::SparseMatrixCSC, p::Real=2)
...
-        # TODO: compute more efficiently using A.nzval directly
-        return opnorm(Array(A), p)
+    elseif m == 1
+        if p == 1
+            return norm(A.nzval, Inf)
+        elseif p == 2
+            return norm(A.nzval, 2)
+        elseif p == Inf
+            return norm(A.nzval, 1)
+        end
```
Comment (dkarrasch) [inline stdlib/SparseArrays/src/linalg.jl]: There's an `else`-branch missing, isn't it? What if `m == 1` but `p` is neither 1, 2, nor `Inf`? I guess it should be either `norm(A.nzval, p/(p-1))`, or `opnorm(A.nzval', p)`.

File: stdlib/SparseArrays/test/sparse.jl [hunk 2]
```diff
@@ -1803,6 +1803,18 @@ end
...
+    colM = sprandn(10, 1, 0.6)
+    @test opnorm(colM, 1) ≈ opnorm(Array(colM), 1)
+    @test opnorm(colM) ≈ opnorm(Array(colM))
+    @test opnorm(colM, Inf) ≈ opnorm(Array(colM), Inf)
+
+    # Test (1 x n) sparse matrix
+    rowM = sprandn(1, 10, 0.6)
+    @test opnorm(rowM, 1) ≈ opnorm(Array(rowM), 1)
+    @test opnorm(rowM) ≈ opnorm(Array(rowM))
+    @test opnorm(rowM, Inf) ≈ opnorm(Array(rowM), Inf)
```
Comment (dkarrasch) [inline stdlib/SparseArrays/test/sparse.jl]: I guess we should also have one test each for the case of `p` some non-standard number (not 1, 2, `Inf`).

File: stdlib/SparseArrays/src/linalg.jl (hunk 1 above)
Comment (goggle) [reply stdlib/SparseArrays/src/linalg.jl]: Oh, that's correct. I assumed that `opnorm` does only support `p = 1, 2, Inf` as the last line of the method suggests:
```
throw(ArgumentError("invalid operator p-norm p=$p. Valid: 1, Inf"))
```
But it used to work for arbitrary values of `p` in the case where `A` is a sparse `m x 1` or `1 x n` matrix, so it should probably be kept this way. I'll see what I can do.

Comment (dkarrasch) [reply stdlib/SparseArrays/src/linalg.jl]: Yes, but that's the case when neither `m == 1` nor `n == 1`, i.e., for actual matrices. If any of the two is 1, then `opnorm(Array(A), p)` is computed, without further checks on `p`. I guess you could have
```julia
elseif m == 1
    norm(A.nzval, p / (p - 1))
elseif n == 1
    norm(A.nzval, p)
else
     # the matrix case ...
```

Comment (goggle) [reply stdlib/SparseArrays/src/linalg.jl]: Actually, my previous statement was wrong: The current implementation of `opnorm` does not support any values of `p` other than `1, 2, Inf`, not even in the case where `A` is a `m x 1` or `1 x n` sparse matrix:
```julia
using LinearAlgebra
using SparseArrays
A = SparseMatrixCSC([1.0 2.0 0.0 0.3])
opnorm(A, 3)
```
gives
```
ERROR: ArgumentError: invalid p-norm p=3. Valid: 1, 2, Inf
Stacktrace:
 [1] opnorm at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/generic.jl:583 [inlined]
 [2] opnorm(::SparseMatrixCSC{Float64,Int64}, ::Int64) at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/SparseArrays/src/linalg.jl:910
 [3] top-level scope at none:0
```

So the question is: Keep it that way (only allowing `p = 1, 2, Inf`) or allowing more general `p` values for the case where `A` is a sparse `m x 1` or `1 x n` matrix?

@dkarrasch What would be the restrictions on `p` with your suggestion `norm(A.nzval, p / (p - 1))`? Simply allowing `p > 0` with special cases for `p = 1` and `p = Inf`?

Comment (dkarrasch) [reply stdlib/SparseArrays/src/linalg.jl]: Ok, I confused myself. My apologies. The reason is that `opnorm(A::AbstractArray{T,2} where T, p::Real)` is not implemented for any other `p` than `1, 2, Inf`, even if `m == 1 || n == 1`. In the latter case, the opnorm reduces to a (dual) vector norm, which technically is defined for other `p`s. I guess that, for consistency, we should leave it at the three cases. However, you will need to check for `p in (1, 2, Inf)` in the line right below this discussion.

Comment (dkarrasch) [inline stdlib/SparseArrays/src/linalg.jl]: If I'm not mistaken, you now need an else-branch which throws the "wrong p" error?

File: stdlib/SparseArrays/src/linalg.jl [hunk 3]
```diff
@@ -1028,9 +1028,16 @@ function opnorm(A::SparseMatrixCSC, p::Real=2)
     m, n = size(A)
     if m == 0 || n == 0 || isempty(A)
         return float(real(zero(eltype(A))))
-    elseif m == 1 || n == 1
-        # TODO: compute more efficiently using A.nzval directly
-        return opnorm(Array(A), p)
+    elseif m == 1
+        if p == 1
+            return norm(A.nzval, Inf)
```
Comment (fredrikekre) [inline stdlib/SparseArrays/src/linalg.jl]:
```suggestion
            return norm(nzvalview(A), Inf)
```

File: stdlib/SparseArrays/src/linalg.jl [hunk 4]
```diff
@@ -1028,9 +1028,16 @@ function opnorm(A::SparseMatrixCSC, p::Real=2)
...
+        elseif p == 2
+            return norm(A.nzval, 2)
```
Comment (fredrikekre) [inline stdlib/SparseArrays/src/linalg.jl]:
```suggestion
            return norm(nzvalview(A), 2)
```

File: stdlib/SparseArrays/src/linalg.jl [hunk 5]
```diff
@@ -1028,9 +1028,16 @@ function opnorm(A::SparseMatrixCSC, p::Real=2)
...
+        elseif p == Inf
+            return norm(A.nzval, 1)
```
Comment (fredrikekre) [inline stdlib/SparseArrays/src/linalg.jl]:
```suggestion
            return norm(nzvalview(A), 1)
```

File: stdlib/SparseArrays/src/linalg.jl [hunk 6]
```diff
@@ -1028,9 +1028,16 @@ function opnorm(A::SparseMatrixCSC, p::Real=2)
...
+        end
+    elseif n == 1 && p in (1, 2, Inf)
+        return norm(A.nzval, p)
```
Comment (fredrikekre) [inline stdlib/SparseArrays/src/linalg.jl]:
```suggestion
        return norm(nzvalview(A), p)
```

File: stdlib/SparseArrays/src/linalg.jl (hunk 1 above)
Comment (goggle) [reply stdlib/SparseArrays/src/linalg.jl]: @dkarrasch It falls through the `if` block and reaches the last line of the method
```julia
throw(ArgumentError("invalid operator p-norm p=$p. Valid: 1, Inf"))
```
which produces the error message for the *wrong* `p`.

**Review by dkarrasch — APPROVED** [review]

**Pull request merged**